File: clients/go/webhook.go
package crucible

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"time"
)

// maxSigCandidates caps how many v1= values we parse and compare. Mirrors the
// gateway verifySignature defense: an attacker cannot force unbounded HMAC
// comparisons by stuffing the header with many candidates.
const maxSigCandidates = 8

// maxHeaderParts caps the total comma-separated segments to prevent unbounded
// iteration over attacker-controlled input before the v1 candidate cap applies.
const maxHeaderParts = 16

// DefaultTolerance is the maximum webhook age accepted by VerifyWebhook.
// Equals the gateway's 5-minute inbound replay window.
const DefaultTolerance = 5 * time.Minute

// SignatureHeader is the HTTP header that carries the t= timestamp and v1= HMAC digest.
const SignatureHeader = "X-Crucible-Signature"

// TimestampHeader is the HTTP header that carries the delivery Unix timestamp.
// It is provided for logging and tracing; do NOT pass it to VerifyWebhook —
// the timestamp used for replay protection is extracted from SignatureHeader.
const TimestampHeader = "X-Crucible-Timestamp"

// WebhookError is returned when X-Crucible-Signature verification fails.
type WebhookError struct {
	msg string
}

func (e *WebhookError) Error() string   { return "crucible webhook: " + e.msg }
func (e *WebhookError) Message() string { return e.msg }

// VerifyWebhook verifies the X-Crucible-Signature on a webhook delivery from the gateway.
// secretHex is the hex-encoded signing secret shown on the dashboard endpoint page.
// sigHeader is the raw value of the X-Crucible-Signature header (format: t=<ts>,v1=<hex>).
// body is the unmodified request body bytes.
// tolerance is the maximum accepted age; pass 0 to use DefaultTolerance.
// All errors are *WebhookError. A non-nil error means the payload must not be trusted.
func VerifyWebhook(secretHex, sigHeader string, body []byte, tolerance time.Duration) error {
	if tolerance == 0 {
		tolerance = DefaultTolerance
	}
	if tolerance < 0 {
		return &WebhookError{"negative tolerance not allowed"}
	}
	if len(secretHex) == 0 || len(secretHex)%2 != 0 {
		return &WebhookError{"invalid secretHex: must be non-empty even-length hex string"}
	}
	secret, hexErr := hex.DecodeString(secretHex)
	if hexErr != nil {
		return &WebhookError{"invalid secretHex: contains non-hex characters"}
	}

	// Capture clock before any attacker-controlled parsing so the sampled instant
	// is not shifted by header-parsing time.
	now := time.Now()

	timestamp, sigs, parseErr := parseSignatureHeader(sigHeader)
	if parseErr != nil {
		return parseErr
	}

	// Bound length to match the TypeScript /^\d{1,15}$/ guard: 15 digits covers
	// all real Unix timestamps and prevents ParseInt from processing monster strings.
	if len(timestamp) > 15 {
		return &WebhookError{"bad timestamp in signature header"}
	}
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return &WebhookError{"bad timestamp in signature header"}
	}
	age := now.Sub(time.Unix(ts, 0))
	if age < 0 {
		return &WebhookError{"webhook timestamp in the future"}
	}
	if age > tolerance {
		return &WebhookError{"webhook timestamp too old (replay protection)"}
	}

	mac := hmac.New(sha256.New, secret)
	// hash.Hash.Write never returns an error; blank identifiers are explicit acknowledgement.
	_, _ = mac.Write([]byte(timestamp))
	_, _ = mac.Write([]byte("."))
	_, _ = mac.Write(body)
	expected := mac.Sum(nil)

	for _, sigHex := range sigs {
		// Length guard mirrors the TypeScript sig.length !== 64 check: reject
		// wrong-length inputs on the fast path before calling the decoder.
		if len(sigHex) != sha256.Size*2 {
			continue
		}
		candidate, hexErr := hex.DecodeString(sigHex)
		if hexErr != nil || len(candidate) != sha256.Size {
			continue
		}
		if hmac.Equal(candidate, expected) {
			return nil
		}
	}
	return &WebhookError{"no matching v1 signature"}
}

func parseSignatureHeader(header string) (string, []string, error) {
	var (
		timestamp string
		sigs      []string
	)
	if header == "" {
		return "", nil, &WebhookError{"missing X-Crucible-Signature header"}
	}
	parts := strings.Split(header, ",")
	if len(parts) > maxHeaderParts {
		return "", nil, &WebhookError{"malformed X-Crucible-Signature header"}
	}
	for _, part := range parts {
		kv := strings.SplitN(part, "=", 2)
		if len(kv) != 2 {
			return "", nil, &WebhookError{"malformed X-Crucible-Signature header"}
		}
		switch kv[0] {
		case "t":
			timestamp = kv[1]
		case "v1":
			if len(sigs) >= maxSigCandidates {
				continue
			}
			sigs = append(sigs, kv[1])
		// Unknown keys (e.g. future v2=) are silently ignored for forward compatibility.
		}
	}
	if timestamp == "" || len(sigs) == 0 {
		return "", nil, &WebhookError{"malformed X-Crucible-Signature header"}
	}
	return timestamp, sigs, nil
}

File: clients/typescript/src/webhook.ts
import { createHmac, timingSafeEqual } from "node:crypto";

// maxSigCandidates caps parsed v1= values to prevent unbounded HMAC comparisons
// on header-stuffed requests. Mirrors gateway verifySignature.
const MAX_SIG_CANDIDATES = 8;

/** Default tolerance in ms: 5 minutes, matching the gateway's inbound replay window. */
export const DEFAULT_TOLERANCE_MS = 5 * 60 * 1000;

/** HTTP header carrying the t= timestamp and v1= HMAC digest. */
export const SIGNATURE_HEADER = "X-Crucible-Signature";

/** HTTP header carrying the delivery Unix timestamp. */
export const TIMESTAMP_HEADER = "X-Crucible-Timestamp";

/** Thrown when X-Crucible-Signature verification fails. */
export class WebhookVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WebhookVerificationError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/**
 * verifyWebhook verifies the X-Crucible-Signature on a webhook delivery from the gateway.
 *
 * @param secretHex - hex-encoded signing secret from the dashboard endpoint page
 * @param sigHeader - raw value of the X-Crucible-Signature header (t=<ts>,v1=<hex>)
 * @param body - unmodified request body as Buffer or string
 * @param toleranceMs - maximum age in ms; defaults to DEFAULT_TOLERANCE_MS (5 min)
 * @throws {WebhookVerificationError} when the signature does not match or is expired
 */
export function verifyWebhook(
  secretHex: string,
  sigHeader: string,
  body: Buffer | string,
  toleranceMs: number = DEFAULT_TOLERANCE_MS,
): void {
  if (toleranceMs < 0) {
    throw new WebhookVerificationError("negative tolerance not allowed");
  }
  if (!secretHex || secretHex.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(secretHex)) {
    throw new WebhookVerificationError(
      "invalid secretHex: must be non-empty even-length hex string",
    );
  }
  const secret = Buffer.from(secretHex, "hex");

  const { timestamp, sigs } = parseSignatureHeader(sigHeader);

  // Validate before parsing: reject whitespace, hex prefixes, and non-digit chars.
  // /^\d+$/ is consistent with Go's strconv.ParseInt strict decimal parsing.
  if (!/^\d+$/.test(timestamp)) {
    throw new WebhookVerificationError("bad timestamp in signature header");
  }
  const ts = parseInt(timestamp, 10);
  // Reject non-finite or non-safe-integer timestamps to prevent overflow in ts*1000.
  if (!Number.isFinite(ts) || !Number.isSafeInteger(ts)) {
    throw new WebhookVerificationError("bad timestamp in signature header");
  }
  const nowMs = Date.now();
  const tsMs = ts * 1000;
  if (tsMs > nowMs) {
    throw new WebhookVerificationError("webhook timestamp in the future");
  }
  if (nowMs - tsMs > toleranceMs) {
    throw new WebhookVerificationError("webhook timestamp too old (replay protection)");
  }

  const bodyBuf = Buffer.isBuffer(body) ? body : Buffer.from(body);
  const mac = createHmac("sha256", secret);
  mac.update(timestamp);
  mac.update(".");
  mac.update(bodyBuf);
  const expected = mac.digest();

  for (const sig of sigs) {
    // Reject wrong-length strings first: a 64-char hex string with a trailing
    // non-hex char still decodes to 32 bytes, bypassing the candidate.length check.
    if (sig.length !== 64) continue;
    const candidate = Buffer.from(sig, "hex");
    // Non-hex chars cause Buffer.from to produce a shorter buffer; timingSafeEqual
    // throws on length mismatch, so we must filter before calling it.
    if (candidate.length !== 32) continue;
    if (timingSafeEqual(candidate, expected)) return;
  }
  throw new WebhookVerificationError("no matching v1 signature");
}

function parseSignatureHeader(header: string): { timestamp: string; sigs: string[] } {
  if (!header) {
    throw new WebhookVerificationError("missing X-Crucible-Signature header");
  }
  let timestamp = "";
  const sigs: string[] = [];
  for (const part of header.split(",")) {
    const idx = part.indexOf("=");
    if (idx < 0) {
      throw new WebhookVerificationError("malformed X-Crucible-Signature header");
    }
    const key = part.slice(0, idx);
    const val = part.slice(idx + 1);
    if (key === "t") {
      timestamp = val;
    } else if (key === "v1") {
      if (sigs.length < MAX_SIG_CANDIDATES) sigs.push(val);
    }
    // Unknown keys (e.g. future v2=) are silently ignored for forward compatibility.
  }
  if (!timestamp || sigs.length === 0) {
    throw new WebhookVerificationError("malformed X-Crucible-Signature header");
  }
  return { timestamp, sigs };
}
